Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean

```diff
@@ -446,4 +450,711 @@ alias nonsingular_of_affine_of_Z_ne_zero := nonsingular_of_Z_ne_zero
+
+end Doubling
+
+section Addition
+
+/-! ### Addition formulae -/
+
+/-- The unit associated to the addition of a non-2-torsion point with its negation.
+More specifically, the unit `u` such that `W.add P Q = u • ![0, 1, 0]` where `P x / P z = Q x / Q z`
+but `P ≠ W.neg P`. -/
+def addU (P Q : Fin 3 → F) : F :=
+  -(P y * Q z - Q y * P z) ^ 3 / (P z * Q z)
+
+lemma addU_smul {P Q : Fin 3 → F} (hPz : P z ≠ 0) (hQz : Q z ≠ 0) {u v : F} (hu : u ≠ 0)
+    (hv : v ≠ 0) : addU (u • P) (v • Q) = (u * v) ^ 2 * addU P Q := by
+  field_simp [addU, smul_fin3_ext]
+  ring1
+
+lemma addU_of_Z_eq_zero_left {P Q : Fin 3 → F} (hPz : P z = 0) : addU P Q = 0 := by
+  rw [addU, hPz, zero_mul, div_zero]
+
+lemma addU_of_Z_eq_zero_right {P Q : Fin 3 → F} (hQz : Q z = 0) : addU P Q = 0 := by
+  rw [addU, hQz, mul_zero <| P z, div_zero]
+
+lemma addU_ne_zero_of_Y_ne {P Q : Fin 3 → F} (hPz : P z ≠ 0) (hQz : Q z ≠ 0)
+    (hy : P y * Q z ≠ Q y * P z) : addU P Q ≠ 0 :=
+  div_ne_zero (neg_ne_zero.mpr <| pow_ne_zero 3 <| sub_ne_zero.mpr hy) <| mul_ne_zero hPz hQz
+
+lemma isUnit_addU_of_Y_ne {P Q : Fin 3 → F} (hPz : P z ≠ 0) (hQz : Q z ≠ 0)
+    (hy : P y * Q z ≠ Q y * P z) : IsUnit (addU P Q) :=
+  (addU_ne_zero_of_Y_ne hPz hQz hy).isUnit
+
+variable (W') in
+/-- The $Z$-coordinate of the addition of two distinct point representatives. -/
+def addZ (P Q : Fin 3 → R) : R :=
+  -3 * P x ^ 2 * Q x * Q z + 3 * P x * Q x ^ 2 * P z + P y ^ 2 * Q z ^ 2 - Q y ^ 2 * P z ^ 2
+    + W'.a₁ * P x * P y * Q z ^ 2 - W'.a₁ * Q x * Q y * P z ^ 2 - W'.a₂ * P x ^ 2 * Q z ^ 2
+    + W'.a₂ * Q x ^ 2 * P z ^ 2 + W'.a₃ * P y * P z * Q z ^ 2 - W'.a₃ * Q y * P z ^ 2 * Q z
+    - W'.a₄ * P x * P z * Q z ^ 2 + W'.a₄ * Q x * P z ^ 2 * Q z
+
+lemma addZ_eq' {P Q : Fin 3 → R} (hP : W'.Equation P) (hQ : W'.Equation Q) :
+    W'.addZ P Q * (P z * Q z) = (P x * Q z - Q x * P z) ^ 3 := by
+  linear_combination (norm := (rw [addZ]; ring1))
+    Q z ^ 3 * (equation_iff _).mp hP - P z ^ 3 * (equation_iff _).mp hQ
+
+lemma addZ_eq {P Q : Fin 3 → F} (hP : W.Equation P) (hQ : W.Equation Q) (hPz : P z ≠ 0)
+    (hQz : Q z ≠ 0) : W.addZ P Q = (P x * Q z - Q x * P z) ^ 3 / (P z * Q z) := by
+  rw [← addZ_eq' hP hQ, mul_div_cancel_right₀ _ <| mul_ne_zero hPz hQz]
+
+lemma addZ_smul (P Q : Fin 3 → R) (u v : R) :
+    W'.addZ (u • P) (v • Q) = (u * v) ^ 2 * W'.addZ P Q := by
+  simp only [addZ, smul_fin3_ext]
+  ring1
+
+lemma addZ_self {P : Fin 3 → R} : W'.addZ P P = 0 := by
+  rw [addZ]
+  ring1
+
+lemma addZ_of_Z_eq_zero_left [NoZeroDivisors R] {P Q : Fin 3 → R} (hP : W'.Equation P)
+    (hPz : P z = 0) : W'.addZ P Q = P y ^ 2 * Q z * Q z := by
+  rw [addZ, hPz, X_eq_zero_of_Z_eq_zero hP hPz]
+  ring1
+
+lemma addZ_of_Z_eq_zero_right [NoZeroDivisors R] {P Q : Fin 3 → R} (hQ : W'.Equation Q)
+    (hQz : Q z = 0) : W'.addZ P Q = -(Q y ^ 2 * P z) * P z := by
+  rw [addZ, hQz, X_eq_zero_of_Z_eq_zero hQ hQz]
+  ring1
+
+lemma addZ_of_X_eq [NoZeroDivisors R] {P Q : Fin 3 → R} (hP : W'.Equation P) (hQ : W'.Equation Q)
```

Comment:
Maybe somewhere (docstring? module docstring?) it should be mentioned that `X_eq` means "X coordinates of P and Q are equal once normalised by dividing by Z coordinates"? Or maybe not -- I guess I just figured this out by looking at the statement.

Reply:
I will remember to add these into the module docstrings when I refactor them into a separate file.